clang-tools-extra/clang-tidy/readability/UseStdMinMaxCheck.cpp

```diff
@@ -0,0 +1,132 @@
+  IncludeInserter.registerPreprocessor(PP);
+}
+
+void UseStdMinMaxCheck::check(const MatchFinder::MatchResult &Result) {
+  const auto *lhsVar1 = Result.Nodes.getNodeAs<Expr>("lhsVar1");
+  const auto *rhsVar1 = Result.Nodes.getNodeAs<Expr>("rhsVar1");
+  const auto *lhsVar2 = Result.Nodes.getNodeAs<Expr>("lhsVar2");
+  const auto *rhsVar2 = Result.Nodes.getNodeAs<Expr>("rhsVar2");
+  const auto *ifStmt = Result.Nodes.getNodeAs<IfStmt>("ifStmt");
+  auto &Context = *Result.Context;
+  const SourceManager &Source = Context.getSourceManager();
+
+  if (!lhsVar1 || !rhsVar1 || !lhsVar2 || !rhsVar2 || !ifStmt)
+    return;
+
+  const auto *binaryOp = dyn_cast<BinaryOperator>(ifStmt->getCond());
+  if (!binaryOp)
+    return;
+
+  SourceLocation ifLocation = ifStmt->getIfLoc();
+  SourceLocation thenLocation = ifStmt->getEndLoc();
+
+  auto lhsVar1Str =
+      Lexer::getSourceText(Source.getExpansionRange(lhsVar1->getSourceRange()),
+                           Context.getSourceManager(), Context.getLangOpts());
+
+  auto lhsVar2Str =
+      Lexer::getSourceText(Source.getExpansionRange(lhsVar2->getSourceRange()),
+                           Context.getSourceManager(), Context.getLangOpts());
+
+  auto rhsVar1Str =
+      Lexer::getSourceText(Source.getExpansionRange(rhsVar1->getSourceRange()),
+                           Context.getSourceManager(), Context.getLangOpts());
+
+  auto replacementMax = lhsVar2Str.str() + " = std::max(" + lhsVar1Str.str() +
+                        ", " + rhsVar1Str.str() + ");";
+  auto replacementMin = lhsVar2Str.str() + " = std::min(" + lhsVar1Str.str() +
+                        ", " + rhsVar1Str.str() + ");";
+  auto *operatorStr = binaryOp->getOpcodeStr().data();
+
+  if (binaryOp->getOpcode() == BO_LT || binaryOp->getOpcode() == BO_LE) {
+    if (tidy::utils::areStatementsIdentical(lhsVar1, lhsVar2, Context) &&
+        tidy::utils::areStatementsIdentical(rhsVar1, rhsVar2, Context)) {
+
+      diag(ifStmt->getIfLoc(), "use `std::max` instead of `%0`")
+          << operatorStr
+          << FixItHint::CreateReplacement(SourceRange(ifLocation, thenLocation),
+                                          std::move(replacementMax))
+          << IncludeInserter.createIncludeInsertion(
+                 Source.getFileID(ifStmt->getBeginLoc()), AlgorithmHeader);
+    } else if (tidy::utils::areStatementsIdentical(lhsVar1, rhsVar2, Context) &&
+               tidy::utils::areStatementsIdentical(rhsVar1, lhsVar2, Context)) {
+      diag(ifStmt->getIfLoc(), "use `std::min` instead of `%0`")
+          << operatorStr
+          << FixItHint::CreateReplacement(SourceRange(ifLocation, thenLocation),
+                                          std::move(replacementMin))
+          << IncludeInserter.createIncludeInsertion(
+                 Source.getFileID(ifStmt->getBeginLoc()), AlgorithmHeader);
+    }
+  } else if (binaryOp->getOpcode() == BO_GT || binaryOp->getOpcode() == BO_GE) {
```

Comment:
merge those 2 min ifs, and 2 max ifs.
And then you will get "condition" that you could translate into AST matchers.